gitlab-pipeline/stage/test.yml

```diff
@@ -202,27 +202,32 @@ test:backend-integration:azblob:enterprise:
     # Dependencies for post job status and io stats
     - apk --update add curl jq sysstat hdparm
 
-    # Start dockerd in the background
-    # The entrypoint script additionally uses docker-init as workarround of known
-    # containerd issue - https://github.com/docker-library/docker/issues/318
-    - export TINI_SUBREAPER=true # https://github.com/krallin/tini#subreaping
-    - /bin/sh -c "/usr/local/bin/dockerd-entrypoint.sh &"
-
     # Make sure docker clients (cli and python) interacts with dockerd via socket
     - unset DOCKER_HOST
     - unset DOCKER_TLS_VERIFY
     - unset DOCKER_CERT_PATH
 
-    # Wait for dockerd to start
+    # Start dockerd in the background
+    - export TINI_SUBREAPER=true # https://github.com/krallin/tini#subreaping
     - |-
-      MAX_WAIT=30
-      while [ $MAX_WAIT -gt 0 ]; do
-        echo "[$(date +%F_%T)] MAX_WAIT=${MAX_WAIT}"; ps afx # Debug information
-        docker version >/dev/null && break
-        MAX_WAIT=$(($MAX_WAIT - 1))
+      # The entrypoint script additionally uses docker-init as workarround of known
+      # containerd issue - https://github.com/docker-library/docker/issues/318
+      run_dockerd() {
+        /bin/sh -c '/usr/local/bin/dockerd-entrypoint.sh &>${CI_PROJECT_DIR}/dockerd.log &'
+      }
+      # Run dockerd and wait it to start
+      MAX_WAIT=60
+      while [ ${MAX_WAIT} -gt 0 ]; do
+        echo "[$(date +%F_%T)] MAX_WAIT=${MAX_WAIT}"; ps # Debug information
+        if docker version &>/dev/null; then
+          docker version # Verify that the dockerd is up and running
+          break
+        elif ! ps -o comm | grep -q -E "docker-init|dockerd|containerd"; then
```

Comment:
This will only restart when all of these processes exit, is that the intention? At first glance I was expecting that if _any_ of the three is not running anymore we should retry.

Reply:
Local testing and logs from today's failed nightlies shows that: if `containerd` "dies" before `dockerd` is started then all the chain of processes terminates; if `containerd` is killed after `dockerd` is started then it restarts; if `dockerd` or `docker-init` is killed then all other processes also terminates. So, I think it's ok. Don't really want to dive into an over complication with killing processes when in this implementation it's possible to catch an intermediate state when it's starting (when it's normal that not all processes are started), especially when can't easily create an other states locally. Please share an ideas if you have something better in mind.

Reply:
Thanks for the experiments and the explanation :+1: